rpc/backend/utils.go

```diff
@@ -365,20 +365,81 @@ func ShouldIgnoreGasUsed(res *abci.ExecTxResult) bool {
 	return res.GetCode() == 11 && strings.Contains(res.GetLog(), "no block gas left to run tx: out of gas")
 }
 
-// GetLogsFromBlockResults returns the list of event logs from the tendermint block result response
+// GetLogsFromBlockResults returns the list of event logs from the tendermint block result response.
+// If needed, it re-indexes logs before returning
 func GetLogsFromBlockResults(blockRes *cmtrpctypes.ResultBlockResults) ([][]*ethtypes.Log, error) {
 	blockLogs := [][]*ethtypes.Log{}
 	for _, txResult := range blockRes.TxsResults {
 		logs, err := AllTxLogsFromEvents(txResult.Events)
 		if err != nil {
 			return nil, err
 		}
-
 		blockLogs = append(blockLogs, logs...)
 	}
+
+	if needsReindexing(blockLogs) {
+		reindexLogs(blockLogs)
+	}
+
 	return blockLogs, nil
 }
 
+// needsReindexing checks if logs in the block need re-indexing.Returns true if:
+//  1. TxIndex values are not strictly increasing across transaction groups
+//  2. Any log Index values are duplicated
+//  3. Log Index values are not monotonically increasing across all groups(transactions)
+func needsReindexing(blockLogs [][]*ethtypes.Log) bool {
+	seenLogIndices := make(map[uint]bool)
+	var lastTxIndex uint
+	var lastLogIndex uint
+	isFirstTx := true
+	isFirstLog := true
+
+	for _, txLogs := range blockLogs {
+		if len(txLogs) == 0 {
+			continue
+		}
+
+		groupTxIndex := txLogs[0].TxIndex
+		if !isFirstTx && groupTxIndex <= lastTxIndex {
+			return true
+		}
+		lastTxIndex = groupTxIndex
+		isFirstTx = false
+
+		for _, entry := range txLogs {
+			if seenLogIndices[entry.Index] {
+				return true
+			}
+			seenLogIndices[entry.Index] = true
+			if !isFirstLog && entry.Index <= lastLogIndex {
+				return true
+			}
+			lastLogIndex = entry.Index
+			isFirstLog = false
+		}
+	}
+	return false
+}
+
+// reindexLogs assigns unique TxIndex and Index values to all logs in the block.
+func reindexLogs(blockLogs [][]*ethtypes.Log) {
```

Comment:
maybe we can just call this every time without needReindexing function  but both works